File: kotlin/0084-largest-rectangle-in-histogram.kt
class Solution {
    fun largestRectangleArea(heights: IntArray): Int {
        val st = LinkedList<IntArray>()
        st.push(intArrayOf(-1, 0))
        var max = 0

        for (i in heights.indices) {
            var start = i

            while (!st.isEmpty() && st.peek()[1] > heights[i]) {
                val curr = st.pop()
                val height = curr[1]
                start = curr[0]
                max = Math.max(max, height * (i - start))
            }

            st.push(intArrayOf(start, heights[i]))
        }

        while (!st.isEmpty()) {
            val curr = st.pop()
            max = Math.max(max, curr[1] * (heights.size - curr[0]))
        }

        return max
    }
}

File: swift/0084-largest-rectangle-in-histogram.swift
/**
 * Question Link: https://leetcode.com/problems/largest-rectangle-in-histogram/
 */

class LargestRectangleInHistogram {
    func largestRectangleArea(_ heights: [Int]) -> Int {
        var maxArea = 0
        var stack = [(Int, Int)]()

        for (i, h) in heights.enumerated() {
            var start = i
            while !stack.isEmpty && stack.last!.1 > h {
                let (index, height) = stack.removeLast()
                maxArea = max(maxArea, height * (i - index))
                start = index
            }
            stack.append((start, h))
        }

        for (i, h) in stack {
            maxArea = max(maxArea, h * (heights.count - i))
        }

        return maxArea
    }
}